src/replication/replication_slave_node.cpp

```diff
@@ -47,72 +48,160 @@ namespace cubreplication
     , m_lc (NULL)
     , m_master_identity ("")
     , m_transfer_receiver (NULL)
+    , m_ctrl_sender_daemon (NULL)
+    , m_ctrl_sender (NULL)
   {
     m_stream = stream;
     m_stream_file = stream_file;
-
-    m_lc = new log_consumer ();
-    m_lc->set_stream (m_stream);
-
-    /* start log_consumer daemons and apply thread pool */
-    m_lc->start_daemons ();
   }
 
   slave_node::~slave_node ()
   {
-    delete m_transfer_receiver;
-    m_transfer_receiver = NULL;
+    stop_and_destroy_online_repl ();
+  }
 
-    delete m_lc;
-    m_lc = NULL;
+  int slave_node::setup_protocol (cubcomm::channel &chn)
+  {
+    UINT64 pos = 0, expected_magic;
+    std::size_t max_len = sizeof (UINT64);
 
-    m_stream_file->remove_sync_notifier ();
+    if (chn.send ((char *) &replication_node::SETUP_REPLICATION_MAGIC, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
 
-    if (m_ctrl_sender != NULL)
+    if (chn.recv ((char *) &expected_magic, max_len) != css_error_code::NO_ERRORS)
       {
-	m_ctrl_sender->stop ();
-	cubthread::get_manager ()->destroy_daemon_without_entry (m_ctrl_sender_daemon);
-	delete m_ctrl_sender;
-	m_ctrl_sender = NULL;
+	return ER_FAILED;
+      }
+
+    if (expected_magic != replication_node::SETUP_REPLICATION_MAGIC)
+      {
+	er_log_debug_replication (ARG_FILE_LINE, "slave_node::setup_protocol error in setup protocol");
+	assert (false);
+	return ER_FAILED;
+      }
+
+    if (chn.recv ((char *) &pos, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
       }
+
+    m_source_available_pos = ntohi64 (pos);
+
+    er_log_debug_replication (ARG_FILE_LINE, "slave_node::setup_protocol available pos :%lld",
+			      m_source_available_pos);
+
+    return NO_ERROR;
+  }
+
+  bool slave_node::need_replication_copy (const cubstream::stream_position start_position) const
+  {
+    if (m_source_available_pos >= start_position
+	&& m_source_available_pos - start_position < ACCEPTABLE_POS_DIFF_BEFORE_COPY)
```

Comment:
What is the purpose of accepting only a difference of ACCEPTABLE_POS_DIFF_BEFORE_COPY?

Reply:
If the difference between slave (recovered) position and master current position is too big, it might be more efficient to start by making a new copy.
This may eventually become a system parameter since it depends on database size and user preference which behavior to have.